plenum/server/replica.py

```diff
@@ -961,19 +966,19 @@ def dispatchThreePhaseMsg(self, msg: ThreePhaseMsg, sender: str) -> Any:
         """
         senderRep = self.generateName(sender, self.instId)
 
-        if self.has_already_ordered(msg.viewNo, msg.ppSeqNo):
-            self.discard(msg, 'already ordered 3 phase message', self.logger.trace)
-            return
-
-        if self.isPpSeqNoBetweenWaterMarks(msg.ppSeqNo):
-            if self.can_pp_seq_no_be_in_view(msg.viewNo, msg.ppSeqNo):
-                self.threePhaseRouter.handleSync((msg, senderRep))
-            else:
-                self.discard(msg, 'un-acceptable pp seq no from previous view', self.logger.warning)
+        result, reason = self.validator.validate_3pc_msg(msg)
+        if result == DISCARD:
+            self.discard(msg, "{} discard message {} from {} "
+                              "with the reason: {}".format(self, msg, sender, reason),
+                         self.logger.trace)
+        elif result == PROCESS:
+            self.threePhaseRouter.handleSync((msg, senderRep))
         else:
-            self.logger.debug("{} stashing 3 phase message {} since ppSeqNo {} is not between {} and {}".
-                              format(self, msg, msg.ppSeqNo, self.h, self.H))
-            self.stashOutsideWatermarks((msg, sender))
+            self.logger.debug("{} stashing 3 phase message {} with "
+                              "the reason: {}".format(self, msg, reason))
+            self.stasher.stash((msg, sender), result)
+            self.stashOutsideWatermarks((msg, sender)) # ?
```

Comment:
I think `stashOutsideWatermarks` needs to be removed at all, since the stasher is now responsible for this.

Reply:
Removed